lab06-kaylee/server.js

```diff
@@ -0,0 +1,58 @@
+'use strict';
+
+const Client = require('./model/client.js');
+const net = require('net');
+const EE = require('events').EventEmitter;
+const ee = new EE();
+const server = module.exports = net.createServer();
+const PORT = process.env.PORT || 3000;
+const pool = [];
+
+ee.on('default', (client, string) => {
+  client.socket.write(`Not a valid command: ${string.split(' ', 1)}\n`);
+});
+
+ee.on('/all', (client, string) => {
+  pool.forEach(c => c.socket.write(`${client.nickName}: ${string}`));
+});
+
+ee.on('/nick', (client, string) => {
+  client.nickName = string.trim();
+  pool.forEach(c => c.socket.write(`You changed your nickname to: ${client.nickName}\n`));
+});
+
+ee.on('/dm', (client, string) => {
+  pool.forEach(receiver => {
+    client.socket.write(`To ${receiver.nickName}: ${string.split(' ').slice(1).join(' ')}`);
+    receiver.socket.write(`From ${client.nickName}: ${string.split(' ').slice(1).join(' ')}`);
+  });
+});
+
+let closeSocket = function(socket) {
+  pool.splice(pool.indexOf(socket), 1);
+  pool.forEach(c => c.socket.write(`${c.userName} has left the chatroom.`));
+};
+
+server.on('connection', socket => {
+  let client = new Client(socket);
+  pool.push(client);
+  pool.forEach(c => c.socket.write(`${client.userName} has connected!\n`));
+
+  socket.on('data', data => {
+    let command = data.toString().split(' ').shift().trim();
+    if(command.startsWith('/')) {
+      ee.emit(command, client, data.toString().split(' ').slice(1).join(' '));
+      return;
+    }
+    ee.emit('default', client, data.toString());
+  });
+
+  socket.on('error', err => {
+    if(err) throw err;
+  });
+
+  socket.on('close', closeSocket());
```

Comment:
**TODO FOR LOST POINTS**

I don't know why, but for some reason, Line 54 causes your server to break down on me when I try to connect to a telnet chatroom. The way you've written it, I think it should logically work, but I can't pinpoint as to why. Things work when I changed it to this:

```
  socket.on('close', function(){
    pool.splice(pool.indexOf(client), 1);
    pool.forEach(c => c.socket.write(`${c.userName} has left the chatroom.`));
  });
```

Again, that's pretty much exactly the same thing as what you had when you abstracted the function out at Line 34, but I can't figure out why that causes it to break. 

I took off points because that was a major bug that broke the server an I was unable to do anything in a chatroom until I rewrote the code on my end. **If you show me it works on your machine locally with this version, then I'll give back those points**.